src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,432 @@
+-module(hpr_packet_reporter).
+
+-behaviour(gen_server).
+
+-include("./include/hpr.hrl").
+-include("./grpc/autogen/server/packet_router_pb.hrl").
+-include("./grpc/autogen/server/config_pb.hrl").
+
+% ------------------------------------------------------------------
+%% API Function Exports
+%% ------------------------------------------------------------------
+-export([
+    start_link/1,
+    init_ets/0,
+    report_packet/2,
+    write_packets/0,
+    upload_packets/1,
+    restart_report_interval/1,
+    stop_report_interval/0
+]).
+
+%% ------------------------------------------------------------------
+%% gen_server Function Exports
+%% ------------------------------------------------------------------
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(SERVER, ?MODULE).
+-define(ETS, hpr_packet_report_ets).
+
+-define(FILE_WRITE_OPTIONS, [write, raw, binary, compressed]).
+%% Up to 5 retries, 1 minute sleep
+-define(RETRY_SLEEP_TIME, 60000).
+-define(AWS_RETRY_OPTIONS, [
+    {retry_options, {exponential_with_jitter, {5, ?RETRY_SLEEP_TIME, ?RETRY_SLEEP_TIME}}}
+]).
+
+-record(state, {
+    aws_client :: aws_client:aws_client(),
+    write_dir :: string(),
+    file_path :: string(),
+    max_file_size :: non_neg_integer(),
+    report_interval :: timer:tref() | undefined,
+    interval_duration :: non_neg_integer() | undefined,
+    upload_window :: non_neg_integer(),
+    upload_window_start_time :: non_neg_integer(),
+    bucket :: binary()
+}).
+
+-type interval_duration_ms() :: non_neg_integer().
+-type timestamp_ms() :: non_neg_integer().
+
+-type packet_reporter_opts() :: #{
+    access_key_id => binary(),
+    secret_access_key => binary(),
+    region => binary(),
+    bucket => binary(),
+    interval_duration => timestamp_ms(),
+    write_dir => string(),
+    max_file_size => integer(),
+    upload_window => timestamp_ms(),
+    localstack_host => binary(),
+    localstack_port => binary()
+}.
+
+%%%===================================================================
+%%% API Function Definitions
+%%%===================================================================
+
+-spec start_link(packet_reporter_opts()) -> any().
+start_link(Args) ->
+    gen_server:start_link({local, ?SERVER}, ?SERVER, Args, []).
+
+-spec init_ets() -> ok.
+init_ets() ->
+    ?ETS = ets:new(?ETS, [named_table, bag, public, {write_concurrency, true}]),
+    ok.
+
+-spec report_packet(Packet :: hpr_packet_up:packet(), PacketRoute :: hpr_route:route()) -> ok.
+report_packet(Packet, PacketRoute) ->
+    EncodedPacket = encode_packet(Packet, PacketRoute),
+    ReportTimestamp = erlang:system_time(millisecond),
+    true = ets:insert(?ETS, {ReportTimestamp, EncodedPacket}),
+    ok.
+
+-spec write_packets() -> ok.
+write_packets() ->
+    gen_server:cast(?SERVER, write).
```

Comment:
Style: should the message match the API function name? It makes it a little easier to find the matching callback.